File: NetManager/MagicKit/NetManager/MessageTool/MessageTool.m


#import "MessageTool.h"


@interface MessageTool ()
+ (MessageTool *)shareInstance;

- (void)buildProcessMessage:(NSString *)message;

- (void)buildProcessMessage:(NSString *)message view:(UIView *)view;

- (void)buildMessage:(NSString *)message isError:(BOOL)yesOrNo;

- (void)buildMessage:(NSString *)message view:(UIView *)view isError:(BOOL)yesOrNo;
@end


@implementation MessageTool

@synthesize showMessage;

+ (MessageTool *)shareInstance {
    static MessageTool *sharedInstance = nil;
    static dispatch_once_t onceToken;
    
    dispatch_once(&onceToken, ^{
        sharedInstance = [[MessageTool alloc] init];
    });
    
    return sharedInstance;
    
}

- (void)showDuration {
    [showMessage hide:YES afterDelay:1.46f];
    
}

- (void)buildMessage:(NSString *)message view:(UIView *)view isError:(BOOL)yesOrNo {
    [showMessage removeFromSuperview];
    showMessage = [[MBProgressHUD alloc] initWithView:view];
    showMessage.labelFont = [UIFont systemFontOfSize:16];
    
    if (!yesOrNo)
        [showMessage setCustomView:[[UIImageView alloc] initWithImage:[UIImage imageNamed:@"success.png"]]];
    else
        [showMessage setCustomView:[[UIImageView alloc] initWithImage:[UIImage imageNamed:@"error.png"]]];
    [showMessage setMode:MBProgressHUDModeCustomView];
    [view addSubview:showMessage];
    showMessage.labelText = message;
    [showMessage show:YES];
    [self performSelectorOnMainThread:@selector(showDuration) withObject:nil waitUntilDone:YES];
}

- (void)buildMessage:(NSString *)message isError:(BOOL)yesOrNo {
    [showMessage removeFromSuperview];
    showMessage = [[MBProgressHUD alloc] initWithView:[self getTopLevelWindow]];
    showMessage.labelFont = [UIFont systemFontOfSize:16];
    
    if (!yesOrNo)
        [showMessage setCustomView:[[UIImageView alloc] initWithImage:[UIImage imageNamed:@"success.png"]]];
    else
        [showMessage setCustomView:[[UIImageView alloc] initWithImage:[UIImage imageNamed:@"error.png"]]];
    [showMessage setMode:MBProgressHUDModeCustomView];
    
    [[self getTopLevelWindow] addSubview:showMessage];
    showMessage.labelText = message;
    [showMessage show:YES];
    [self performSelectorOnMainThread:@selector(showDuration) withObject:nil waitUntilDone:YES];
}

- (UIWindow *)getTopLevelWindow {
    UIWindow *window = nil;
    for (UIWindow *_window in [[UIApplication sharedApplication] windows]) {
        if (window == nil) {
            window = _window;
        }
        if (_window.windowLevel > window.windowLevel) {
            window = _window;
        }
    }
    
    return window;
}

- (void)buildProcessMessage:(NSString *)message {
    [showMessage removeFromSuperview];
    showMessage = [[MBProgressHUD alloc] initWithView:[self getTopLevelWindow]];
    [[self getTopLevelWindow] addSubview:showMessage];
    showMessage.labelText = message;
    [showMessage show:YES];
}

- (void)buildProcessMessage:(NSString *)message view:(UIView *)view {
    [showMessage removeFromSuperview];
    showMessage = [[MBProgressHUD alloc] initWithView:view];
    [view addSubview:showMessage];
    showMessage.labelText = message;
    [showMessage show:YES];
}

+ (void)showMessage:(NSString *)message view:(UIView *)view isError:(BOOL)yesOrNo {
    MessageTool *sharedInstance = [MessageTool shareInstance];
    [sharedInstance buildMessage:message view:view isError:yesOrNo];
}

+ (void)showMessage:(NSString *)message isError:(BOOL)yesOrNo {
    MessageTool *sharedInstance = [MessageTool shareInstance];
    [sharedInstance buildMessage:message isError:yesOrNo];
}

+ (MBProgressHUD *)showProcessMessage:(NSString *)message {
    MessageTool *sharedInstance = [MessageTool shareInstance];
    [sharedInstance buildProcessMessage:message];
    return sharedInstance.showMessage;
}

+ (MBProgressHUD *)showProcessMessage:(NSString *)message view:(UIView *)view {
    MessageTool *sharedInstance = [MessageTool shareInstance];
    [sharedInstance buildProcessMessage:message view:view];
    return sharedInstance.showMessage;
}

@end

File: NetManager/MagicKit/NetManager/MessageTool/MessageTool.swift
 

import UIKit

class MessageTool: NSObject, MBProgressHUDDelegate {

    private var showMessage: MBProgressHUD = MBProgressHUD()
    
    
    //MARK: -- Private
    //单例
    class var shareInstance: MessageTool {
        struct Static {
            static var onceToken: dispatch_once_t = 0
            static var instance: MessageTool? = nil
        }
        dispatch_once(&Static.onceToken) {
            Static.instance = MessageTool()
        }
        return Static.instance!
    }
    
    private func showDuration() {
        showMessage.hide(true, afterDelay: 1.46)
    }
    
    private func buildMessage(message: String, view: UIView, isError: Bool) {
        showMessage.removeFromSuperview()
        showMessage = MBProgressHUD(view: view)
        showMessage.labelFont = UIFont.systemFontOfSize(16)
        
        if !isError{
            showMessage.customView = UIImageView(image: UIImage(named: "success.png"))
        }else{
            showMessage.customView = UIImageView(image: UIImage(named: "error.png"))
        }
        showMessage.labelText = message
        showMessage.mode = MBProgressHUDMode.CustomView
        view.addSubview(showMessage)
        showMessage.show(true)
        
        self.performSelectorOnMainThread("showDuration", withObject: nil, waitUntilDone: false)
    }
    
    private func buildMessage(message: String, isError: Bool) {
        showMessage.removeFromSuperview()
        showMessage = MBProgressHUD(view: self.getTopLevelWindow())
        showMessage.labelFont = UIFont.systemFontOfSize(16)
        
        if !isError{
            showMessage.customView = UIImageView(image: UIImage(named: "success.png"))
        }else{
            showMessage.customView = UIImageView(image: UIImage(named: "error.png"))
        }
        showMessage.labelText = message
        showMessage.mode = MBProgressHUDMode.CustomView
        self.getTopLevelWindow().addSubview(showMessage)
        showMessage.show(true)
        
        self.performSelectorOnMainThread("showDuration", withObject: nil, waitUntilDone: false)
    }
    
    private func getTopLevelWindow() -> UIWindow {
        var window : UIWindow = UIWindow()
        for levelWindow in UIApplication.sharedApplication().windows {
            if levelWindow.windowLevel > window.windowLevel {
                window = levelWindow
            }
        }
        return window
    }
    
    private func buildProcessMessage(message: String) {
        showMessage.removeFromSuperview()
        showMessage = MBProgressHUD(view: self.getTopLevelWindow())
        self.getTopLevelWindow().addSubview(showMessage)
        showMessage.labelText = message
        showMessage.show(true)
    }
    
    private func buildProcessMessage(message: String, view: UIView) {
        showMessage.removeFromSuperview()
        showMessage = MBProgressHUD(view: view)
        view.addSubview(showMessage)
        showMessage.labelText = message
        showMessage.show(true)
    }
    
    //MARK: -- Public
    static func showMessage(message: String, isError: Bool) {
        MessageTool.shareInstance.buildMessage(message, isError: isError)
    }
    
    static func showMessage(message: String, view: UIView, isError: Bool) {
        MessageTool.shareInstance.buildMessage(message, view: view, isError: isError)
    }
    
    static func showProcessMessage(message: String) -> MBProgressHUD {
        let tool = MessageTool.shareInstance
        tool.buildProcessMessage(message)
        return tool.showMessage
    }
    
    static func showProcessMessage(message: String, view: UIView) -> MBProgressHUD {
        let tool = MessageTool.shareInstance
        tool.buildProcessMessage(message, view: view)
        return tool.showMessage
    }
    
}
